src/vos/vos_obj.c

```diff
@@ -1928,26 +1938,45 @@ vos_obj_iter_pre_aggregate(daos_handle_t ih)
 	krec = rbund.rb_krec;
 	umm = vos_obj2umm(oiter->it_obj);
 
-	if (!vos_ilog_is_punched(vos_cont2hdl(obj->obj_cont), &krec->kr_ilog, &oiter->it_epr,
-				 &oiter->it_punched, &oiter->it_ilog_info))
+	punched = vos_ilog_is_punched(vos_cont2hdl(obj->obj_cont), &krec->kr_ilog, &oiter->it_epr,
+				      &oiter->it_punched, &oiter->it_ilog_info);
+
+	bmap = krec->kr_bmap;
+
+	if (bmap & KREC_BF_AGG_OPT && !oiter->it_iter.it_for_discard) {
```

Comment:
Will it cleaner and easier to follow if we do the feat flags check in need_aggregate()?

Reply:
Probably.  I think it probably needs a bit of refactoring.   It's tricky to do this because the data is embedded in the tree root/key record and we don't export this via the iterator.